### PR TITLE
dart: Respect LSP binary settings

### PR DESCRIPTION
**Pull request by ybbond**
Enable configuring Dart's LSP from other means of installation types.

Some users don't install the `dart` binary, but uses version manager.

In the example, I uses [FVM](https://fvm.app/) (short for "Flutter Version Manager").

I have tested this with "Install Dev Extensions".

Release Notes:

- N/A

cc other maintainer: @agent3bood @flo80

### CODE REVIEW

**Comment by ybbond** [comment]
@cla-bot check

**Review by maxdeviant — CHANGES_REQUESTED** [review]
What is `fvm`? Is it a different language server?

If so, we’d want to structure them as two different language servers within the extension that the user can then switch between.

**Comment by ybbond** [comment]
> @maxdeviant What is `fvm`? Is it a different language server?
> 
> If so, we’d want to structure them as two different language servers within the extension that the user can then switch between.

It is shorts for "Flutter Version Manager", and every build of Flutter includes Dart. It can detect if fvm config is configured on per-project basis and set specific dart-flutter version for that directory.

I think it's common use case for people to use FVM even if they only use Dart.

I also updated the description to include summary of fvm

File: extensions/dart/extension.toml
Comment (maxdeviant) [inline extensions/dart/extension.toml:4]:
```suggestion
version = "0.0.3"
```

**Comment by medz** [comment]
Do not merge this PR! Please continue to use Dart & Flutter on your computer!!!

FVM will destroy user features! For example, if I install Dart 3.5.x on my computer and enable macros, and FVM installs a copy of Flutter + Dart SDK on its own. This will destroy the SDK I expect to run!

Please stop this mindless behavior!!!

**Comment by medz** [comment]
This PR is as stupid as Bun.js adding CNPM.

Hopefully it won't be merged, or will follow the community's objection like Bun.js did.

**Comment by ybbond** [comment]
whoa, easy there @medz 

If you read the [diff](https://github.com/zed-industries/zed/pull/17494/files), you will notice that this PR will not add `fvm` to your computer.

I'd like to make sure whether you misunderstood the way version managers (in this case `fvm`) works. Because in your deleted reply, you mentioned

> @medz Do not merge this PR! Please continue to use Dart & Flutter on your computer!!!
> 
> FVM will destroy user features! For example, if I install Dart 3.5.x on my computer and enable macros, and FVM installs a copy of Flutter + Dart SDK on its own. This will destroy the SDK I expect to run!
> 
> Please stop this mindless behavior!!!

I'm sorry if this unfortunate behavior ever happened to you, but I think what you can do is to **disable** fvm's global versioning of Flutter / Dart by doing:
```sh
fvm global --unlink
```

Because I think your computer is confusing between independently installed `dart` binary with `fvm`.

---

But, thanks to you @medz. Your written concern made me realize that checking both `dart` and `fvm` —even if for convenience— will not please everybody. I think I should find a way to make this behavior configurable.

That means I need to deepen my knowledge in Rust

**Comment by ezerinz** [comment]
Hey, I don't understand rust much. But doesn't this code prioritize global installed dart over the fvm one?

**Comment by medz** [comment]
> whoa, easy there @medz
> 
> If you read the [diff](https://github.com/zed-industries/zed/pull/17494/files), you will notice that this PR will not add `fvm` to your computer.
> 
> I'd like to make sure whether you misunderstood the way version managers (in this case `fvm`) works. Because in your deleted reply, you mentioned
> 
> > @medz Do not merge this PR! Please continue to use Dart & Flutter on your computer!!!
> > FVM will destroy user features! For example, if I install Dart 3.5.x on my computer and enable macros, and FVM installs a copy of Flutter + Dart SDK on its own. This will destroy the SDK I expect to run!
> > Please stop this mindless behavior!!!
> 
> I'm sorry if this unfortunate behavior ever happened to you, but I think what you can do is to **disable** fvm's global versioning of Flutter / Dart by doing:
> 
> ```shell
> fvm global --unlink
> ```
> 
> Because I think your computer is confusing between independently installed `dart` binary with `fvm`.
> 
> But, thanks to you @medz. Your written concern made me realize that checking both `dart` and `fvm` —even if for convenience— will not please everybody. I think I should find a way to make this behavior configurable.
> 
> That means I need to deepen my knowledge in Rust

@ybbond  Thanks for the clarification, yes I thought it was adding FVM at first, then I noticed the diff, so I deleted the comment (I think I was too aggressive and inappropriate)

But I can't help but think it's stupid to add fvm check, and even I think fvm itself is quite unacceptable to me, if it's dart version manager, I think it's reasonable.

fvm is far less convenient and reasonable than `flutter channel --version <xxx>`. I haven't used fvm for a long time, because fvm makes me feel stupid. (Sorry for the digression)

I think the zed dart extension should be given a setting to allow users to switch Dart SDK instead of going through fvm check. By default, the dark copy installed by the user should be used.

This kind of problem is a bit similar to the CNPM support that Bun.js had a serious dispute over some time ago. In the end, it was designed as universal support, so that all regions with restricted registry access in the world can benefit.

I think the same should be done for the Dart SDK path, rather than making unique optimizations for a specific third-party software (which makes users feel that a certain software is officially supported, which is offensive)

For example, the dart path can be set as an array, and then you can click on the bottom status bar to quickly switch, which will be more popular.

**Comment by ybbond** [comment]
marking this as draft, I will contemplate a little on better way to make this configurable

**Comment by agent3bood** [comment]
Hi @ybbond 
Wile this is a nice feature and many Flutter devs will love me (including myself), I think the Zed should decide about adding 'version managers', take for example do we need to support 'dvm' and 'asdf' which both are applicable to this plugin.

**Comment by agent3bood** [comment]
And we hav another version manager, >>> https://puro.dev/

I think it is better to make the language server path configurable by Zed itself.

**Comment by ybbond** [comment]
> @agent3bood: I think it is better to make the language server path configurable by Zed itself.

yes I agree with this

**Comment by ybbond** [comment]
hi @agent3bood @maxdeviant . I have pushed new implementation which allow various `dart` installation methods. I use the built-in Zed's lsp settings to make it working.

I also added documentation and edited the description.

Please recheck 😄

**Review by maxdeviant — APPROVED** [review]
I think respecting the LSP binary settings makes sense.

Thanks!

**Pull request merged**
